src/main/java/org/sonatype/nexus/ci/nxrm/v3/ComponentUploaderNxrm3.groovy

```diff
@@ -0,0 +1,94 @@
+import org.sonatype.nexus.ci.nxrm.ComponentUploader
+import org.sonatype.nexus.ci.nxrm.MavenCoordinate
+
+import groovy.transform.InheritConstructors
+import hudson.model.Result
+
+import static org.sonatype.nexus.ci.util.RepositoryManagerClientUtil.nexus3Client
+
+@SuppressWarnings(['CatchException', 'AbcMetric', 'MethodSize'])
+@InheritConstructors
+class ComponentUploaderNxrm3
+    extends ComponentUploader
+{
+  @Override
+  void upload(final Map<MavenCoordinate, List<RemoteMavenAsset>> remoteMavenComponents,
+                        String nxrmRepositoryId)
+  {
+    def nxrmClient = getRepositoryManagerClient(nxrmConfiguration)
+
+    remoteMavenComponents.each { mavenCoordinate, remoteMavenAssets ->
+      try {
+        logger.println("Uploading Maven asset with groupId: ${mavenCoordinate.groupId} " +
+            "artifactId: ${mavenCoordinate.artifactId} version: ${mavenCoordinate.version} " +
+            "To repository: ${nxrmRepositoryId}")
+
+
+        def mavenComponentBuilder = MavenComponentBuilder.create()
+            .withGroupId(envVars.expand(mavenCoordinate.groupId))
+            .withArtifactId(envVars.expand(mavenCoordinate.artifactId))
+            .withVersion(envVars.expand(mavenCoordinate.version))
+            .withPackaging(envVars.expand(mavenCoordinate.packaging))
+
+        remoteMavenAssets.eachWithIndex { remoteMavenAsset, idx ->
+          def asset = new DefaultAsset("asset${idx}", remoteMavenAsset.RemotePath.getRemote(),
+              remoteMavenAsset.RemotePath.read())
+
+          mavenComponentBuilder.withAsset(asset, envVars.expand(remoteMavenAsset.Asset.extension),
+              envVars.expand(remoteMavenAsset.Asset.classifier))
+        }
+
+        try {
+          nxrmClient.upload(nxrmRepositoryId, mavenComponentBuilder.build())
+        }
+        catch (RepositoryManagerException ex) {
+          throw new IOException(ex)
+        }
+      }
+      catch (IOException ex) {
+        final String uploadFailed = 'Upload of maven component with GAV ' +
+            "[${mavenCoordinate.groupId}:${mavenCoordinate.artifactId}:${mavenCoordinate.version}] failed"
+
+        logger.println(uploadFailed)
+        logger.println('Failing build due to failure to upload file to Nexus Repository Manager Publisher')
+        run.setResult(Result.FAILURE)
+        throw new IOException(uploadFailed, ex)
+      }
+    }
+
+    logger.println('Successfully Uploaded Maven Assets')
+  }
+
+  RepositoryManagerV3Client getRepositoryManagerClient(final Nxrm3Configuration nxrm3Configuration) {
+    try {
+      nexus3Client(nxrm3Configuration.serverUrl, nxrm3Configuration.credentialsId, nxrm3Configuration.anonymousAccess)
+    }
+    catch (Exception e) {
+      def message = 'Error creating RepositoryManagerClient'
+      logger.println(message)
+      logger.println('Failing build due to error creating RepositoryManagerClient')
+      run.setResult(Result.FAILURE)
+      throw e
```

Comment:
Same question as earlier, would it be better to throw a new exception with your message having the original exception as the cause?